compiler/rustc_hir_typeck/src/check.rs

```diff
@@ -32,18 +33,20 @@ pub(super) fn check_fn<'a, 'tcx>(
 ) -> (FnCtxt<'a, 'tcx>, Option<GeneratorTypes<'tcx>>) {
     // Create the function context. This is either derived from scratch or,
     // in the case of closures, based on the outer context.
-    let mut fcx = FnCtxt::new(inherited, param_env, body.value.hir_id);
+    let def_id = inherited.tcx.hir().opt_local_def_id(body.value.hir_id).unwrap_or(CRATE_DEF_ID);
```

Comment:
This is wrong. A body's `HirId` never has a `LocalDefId`, so this line always returns `CRATE_DEF_ID`, which is not whar we want. (Uses of `opt_local_def_id` are almost always wrong.)

```suggestion
    let def_id = inherited.tcx.hir().body_owner_def_id(body.id());
```

Reply:
Mh, I was thinking that this was too simple, I ca know why the body does not have the `LocalDefId`?

Reply:
For instance, the `LocalDefId` identifies the function item, which is a `hir::Item`. That item *contains* a `BodyId`.
The `BodyId` points to a `Body` which wraps a `Expr` (the indirection is there to help incremental compilation, when looking at HIR of the function signature only).
By convention, the `HirId` inside the `BodyId` is the `HirId` of the `Expr` (`Body::value` is an `Expr`, and you may have seen `body.value.hir_id` plenty of times).

Likewise for closures: the closure *itself* has both a `HirId` and a `LocalDefId`, but the returned expression is the one with the `BodyId`.

This also explains the naming for `body_owner_def_id`: we go one level higher (hence the "owner") when going from the `BodyId` to the `LocalDefId`.

I've been wanting to make `BodyId` store the `LocalDefId` of the function instead, but putting off that work. Your PR will make that significantly easier.